scripts/lib/installer-win32-gen.tsx

```diff
@@ -0,0 +1,238 @@
+          result += `<![CDATA[${ child }]]>`;
+        } else if (child instanceof Element) {
+          result += child.toXML(indent + 2);
+        } else {
+          throw new TypeError(`Don't know how to serialize ${ child } (type ${ typeof child })`);
+        }
+      }
+      if (this.children.some(c => c instanceof Element)) {
+        result += indentString;
+      }
+      result += `</${ this.tag }>${ '\n' }`;
+    }
+
+    return result;
+  }
+}
+
+// When rendering, the JSX tag name is supposed to be the name of a component
+// that's passed to the first argument of React.createElement; so we need plain
+// constants for every element we use.
+const Component = 'Component';
+const ComponentGroup = 'ComponentGroup';
+const ComponentGroupRef = 'ComponentGroupRef';
+const Directory = 'Directory';
+const File = 'File';
+const Fragment = 'Fragment';
+const Shortcut = 'Shortcut';
+const ShortcutProperty = 'ShortcutProperty';
+
+/**
+ * A structure representing the files (but not subdirectories) within a
+ * directory.
+ */
+type directory = {
+  /** The identifier for this directory. */
+  id: string;
+  /** The name of this directory, as the path relative to appDir. */
+  name: string;
+  /** Child directories. */
+  directories: directory[];
+  /** The regular files within this direcotry */
+  files: { name: string, id: string }[];
+};
+
+/** Walk the given directory, determining what files exist. */
+function walk(root: string): Promise<directory> {
+  async function walkDirectory(dir: string): Promise<directory> {
+    const relPath = path.relative(root, dir);
+    const files: { name: string, id: string }[] = [];
+    const result: directory = {
+      id:          '', // Will be updated later
+      name:        relPath,
+      directories: [],
+      files:       [],
+    };
+    const hasher = crypto.createHash('sha256');
+    const children = await fs.promises.readdir(dir, { withFileTypes: true });
+
+    hasher.update(relPath);
+    await Promise.all(children.sort((a, b) => a.name.localeCompare(b.name)).map(async(child) => {
+      if (child.isDirectory()) {
+        result.directories.push(await walkDirectory(path.join(dir, child.name)));
+      } else if (child.isFile()) {
+        const info = await fs.promises.stat(path.join(dir, child.name));
+        const input = `${ child.name }::${ info.size }@${ info.mtimeMs }`;
+        const id = `f${ hasher.copy().update(input).digest('base64url').replaceAll('-', '.') }`;
+
+        files.push({ name: child.name, id });
+      } else {
+        throw new Error(`Could not handle non-regular file ${ path.join(dir, child.name) }`);
+      }
+    }));
+
+    result.directories.sort((a, b) => a.name.localeCompare(b.name));
+
+    files.sort((a, b) => a.name.localeCompare(b.name));
+    result.files = files;
+
+    files.forEach(f => hasher.update(f.id));
+    result.id = `d_${ hasher.digest('base64url').replaceAll('-', '.') }`;
```

Comment:
Per our offline conversation with Mark, should this be consistent with the id on line 115`?

Reply:
Yep, added the underscore (`f_`) on line 115 to match.